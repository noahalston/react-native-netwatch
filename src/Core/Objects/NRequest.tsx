import { ILog, LogType } from '../../types';

export class NRequest implements ILog {
  _id: number = -1;
  type: LogType = 'NR';
  startTime: number = Date.now();
  readyState: number = 0;
  url: string = '';
  shortUrl?: string = '';
  method: string = '';
  status: number = -1;
  endTime: number = 0;
  timeout?: number;
  dataSent: string = '';
  requestHeaders?: any;
  responseHeaders?: any;
  responseContentType?: string;
  responseSize?: number;
  responseType?: string;
  responseURL?: string;
  response: string = '';

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default NRequest;
