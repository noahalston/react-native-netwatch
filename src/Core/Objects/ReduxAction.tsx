import { ILog, LogType } from '../../types';

interface IAction {
  type: string;
  payload: any;
}

export class ReduxAction implements ILog {
  _id: number = -1;
  startTime: number = Date.now();
  type: LogType = 'REDUX';
  action: IAction = { type: '__ERROR:UNDEFINED__', payload: '' };

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default ReduxAction;
