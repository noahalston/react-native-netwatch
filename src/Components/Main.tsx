import * as React from 'react';
import { useState, useEffect } from 'react';
import { View, FlatList, StyleSheet, Modal } from 'react-native';
import {
  Appbar,
  Searchbar,
  Surface,
  IconButton,
  Dialog,
  Paragraph,
  Button,
} from 'react-native-paper';
import Item from './Item';
import ReduxItem from './ReduxItem';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { Settings } from './Settings';
import { ILog, SourceType, RequestMethod, EnumSourceType, EnumFilterType } from '../types';
import RNRequest from '../Core/Objects/RNRequest';
import ReduxAction from '../Core/Objects/ReduxAction';

export interface IProps {
  testId?: string;
  onPress: Function;
  onPressClose: (value: boolean) => void;
  onPressDetail: (value: boolean) => void;
  visible?: boolean;
  reduxActions: ReduxAction[];
  rnRequests: RNRequest[];
  clearAll: Function;
  maxRequests?: number;
}

let reduxList: ReduxAction[] = [];
let rnList: RNRequest[] = [];
// let nList: NRequest[] = [];

export const Main = (props: IProps) => {
  const [requests, setRequests] = useState<Array<ILog>>([]);
  const [searchQuery, setSearchQuery] = useState('');
  const [source, setSource] = useState<SourceType | EnumSourceType>(EnumSourceType.All);
  const [filter, setFilter] = useState<RequestMethod | EnumFilterType>(EnumFilterType.All);
  const [settingsVisible, setSettingsVisible] = useState(false);
  const [deleteVisible, setDeleteVisible] = useState(false);
  const onChangeSearch = (query: string) => setSearchQuery(query);
  const showDialog = () => setSettingsVisible(true);
  const hideSettingsDialog = () => setSettingsVisible(false);
  const showDialogDelete = () => setDeleteVisible(true);
  const hideDialogDelete = () => setDeleteVisible(false);
  const hideDialogAndDelete = () => {
    clearList();
    hideDialogDelete();
  };

  useEffect(() => {
    if (source === EnumSourceType.Redux) {
      if (filter !== EnumFilterType.All) setFilter(EnumFilterType.All);
      reduxList = [...props.reduxActions];
      setRequests(reduxList);
    } else if (source === EnumSourceType.ReactNativeRequest) {
      rnList = [...props.rnRequests];
      setRequests(rnList);
    } else if (source === EnumSourceType.Nativerequest) {
      setRequests([]);
      // nList = [...props.nRequests.reverse()].slice(0, maxRequests);
      // setRequests(nList);
    } else {
      setRequests(mergeArrays(props.reduxActions, props.rnRequests).sort(compare).reverse());
    }
    return () => {};
  }, [props.rnRequests, props.reduxActions]);

  const mergeArrays = (...arrays: Array<ILog[]>) => {
    return [...arrays.flat()];
  };

  const compare = (a: ILog, b: ILog) => {
    const startTimeA = a.startTime;
    const startTimeB = b.startTime;

    let comparison = 0;
    if (startTimeA > startTimeB) {
      comparison = 1;
    } else if (startTimeA < startTimeB) {
      comparison = -1;
    }
    return comparison;
  };

  const filteredRequests = requests.slice(0, props.maxRequests).filter((request) => {
    if (filter === EnumFilterType.All) {
      if (request instanceof ReduxAction) {
        return request.action.type.toLowerCase().includes(searchQuery.toLowerCase());
      }
    }
    // TODO: Add  || request instanceof NRequest when Native request will be available
    if (request instanceof RNRequest) {
      if (filter === request.method || filter === EnumFilterType.All) {
        return (
          request.url?.toLowerCase().includes(searchQuery.toLowerCase()) ||
          request.status === parseInt(searchQuery) ||
          request.method.toLowerCase() === searchQuery.toLowerCase()
        );
      }
    }
    return;
  });

  const clearList = () => {
    props.clearAll();
    setRequests([]);
  };

  return (
    <>
      <Appbar.Header style={styles.header}>
        <IconButton color="white" icon="close" onPress={() => props.onPressClose(false)} />
        <Appbar.Content title="Netwatch" />
      </Appbar.Header>
      <Surface style={styles.surface}>
        <Searchbar
          placeholderTextColor="white"
          iconColor="white"
          inputStyle={styles.searchBarInputStyle}
          style={styles.searchBar}
          placeholder={'Search'}
          onChangeText={onChangeSearch}
          value={searchQuery}
        />
        <IconButton
          // style={{ marginRight: -8 }}
          icon={() => <Icon color="white" size={24} name="delete" />}
          size={24}
          onPress={showDialogDelete}
        />
        <IconButton
          style={{ marginRight: -8 }}
          icon={() => <Icon color="white" size={24} name="settings" />}
          size={24}
          onPress={showDialog}
        />
      </Surface>

      <Settings
        visible={settingsVisible}
        onDismiss={hideSettingsDialog}
        source={source}
        onSetSource={setSource}
        filter={filter}
        onSetFilter={setFilter}
        onPressClear={clearList}
      />

      <Modal animationType="fade" transparent visible={deleteVisible}>
        <Dialog visible={deleteVisible} onDismiss={hideDialogDelete}>
          <Dialog.Title>WARNING</Dialog.Title>
          <Dialog.Content>
            <Paragraph>Do you really want clear the call list?</Paragraph>
          </Dialog.Content>
          <Dialog.Actions>
            <Button onPress={hideDialogDelete}>CANCEL</Button>
            <Button onPress={hideDialogAndDelete}>OK</Button>
          </Dialog.Actions>
        </Dialog>
      </Modal>

      <View style={styles.centeredView}>
        <View style={styles.modalView}>
          <FlatList
            style={styles.flatList}
            keyExtractor={(item) => `${item._id.toString()}${item.startTime}`}
            data={filteredRequests}
            renderItem={({ item }) => {
              if (item instanceof ReduxAction) {
                return <ReduxItem item={item} />;
              } else if (item instanceof RNRequest) {
                return (
                  <Item
                    item={item}
                    onPress={() => {
                      props.onPress(item);
                      props.onPressDetail(true);
                    }}
                  />
                );
              }
              return null;
            }}
          />
        </View>
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  header: {
    backgroundColor: '#212121',
  },
  searchBar: {
    flex: 1,
    backgroundColor: '#616161',
    color: 'white',
  },
  searchBarInputStyle: {
    color: 'white',
  },
  surface: {
    padding: 16,
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#212121',
  },
  flatList: {
    width: '100%',
    backgroundColor: '#bdbdbd',
    padding: 0,
    margin: 0,
  },
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: '100%',
  },
  modalView: {
    width: '100%',
    height: '100%',
  },
});
