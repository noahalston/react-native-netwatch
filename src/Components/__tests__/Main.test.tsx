import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Main, IProps } from '../Main';
import ReduxAction from '../../Core/Objects/ReduxAction';
import { RNRequest } from '../../Core/Objects/RNRequest';

describe('Main test suite', () => {
  let component: ShallowWrapper;
  let props: IProps;
  const onPress = jest.fn();
  const onPressClose = jest.fn();
  const onPressDetail = jest.fn();
  const clearAll = jest.fn();

  it('should render properly', () => {
    givenProps();
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  // GIVEN
  function givenComponent() {
    component = shallow(<Main {...props} />);
  }

  function givenProps() {
    props = {
      onPress,
      onPressClose,
      onPressDetail,
      reduxActions: mockActions,
      rnRequests: mockRequests,
      clearAll,
    };
  }

  const mockActions: ReduxAction[] = [
    {
      _id: 73,
      startTime: 100,
      type: 'REDUX',
      action: { type: '__ERROR:UNDEFINED__', action: '' },
    },
  ];

  const mockRequests: RNRequest[] = [
    {
      _id: 73,
      dataSent: 'dataSent',
      endTime: 1613477575757,
      method: 'GET',
      readyState: 4,
      response: 'response',
      responseContentType: 'application/json',
      responseSize: 0,
      responseType: 'blob',
      responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
      startTime: 1613477574742,
      status: 200,
      timeout: 0,
      type: 'RNR',
      url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
    },
  ];
});
