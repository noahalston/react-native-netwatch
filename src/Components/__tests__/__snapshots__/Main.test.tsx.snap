// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main test suite should render properly 1`] = `
<Fragment>
  <ForwardRef(withTheme(Appbar.Header))
    style={
      Object {
        "backgroundColor": "#212121",
      }
    }
  >
    <ForwardRef(withTheme(IconButton))
      color="white"
      icon="close"
      onPress={[Function]}
    />
    <ForwardRef(withTheme(Appbar.Content))
      title="Netwatch"
    />
  </ForwardRef(withTheme(Appbar.Header))>
  <ForwardRef(withTheme(Surface))
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#212121",
        "flexDirection": "row",
        "padding": 16,
      }
    }
  >
    <ForwardRef(withTheme(undefined))
      iconColor="white"
      inputStyle={
        Object {
          "color": "white",
        }
      }
      onChangeText={[Function]}
      placeholder="Search"
      placeholderTextColor="white"
      style={
        Object {
          "backgroundColor": "#616161",
          "color": "white",
          "flex": 1,
        }
      }
      value=""
    />
    <ForwardRef(withTheme(IconButton))
      icon={[Function]}
      onPress={[Function]}
      size={24}
    />
    <ForwardRef(withTheme(IconButton))
      icon={[Function]}
      onPress={[Function]}
      size={24}
      style={
        Object {
          "marginRight": -8,
        }
      }
    />
  </ForwardRef(withTheme(Surface))>
  <Settings
    filter="ALL"
    onDismiss={[Function]}
    onPressClear={[Function]}
    onSetFilter={[Function]}
    onSetSource={[Function]}
    source="ALL"
    visible={false}
  />
  <Component
    animationType="fade"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <ForwardRef(withTheme(Dialog))
      onDismiss={[Function]}
      visible={false}
    >
      <ForwardRef(withTheme(Dialog.Title))>
        WARNING
      </ForwardRef(withTheme(Dialog.Title))>
      <Dialog.Content>
        <Paragraph>
          Do you really want clear the call list?
        </Paragraph>
      </Dialog.Content>
      <Dialog.Actions>
        <ForwardRef(withTheme(Button))
          onPress={[Function]}
        >
          CANCEL
        </ForwardRef(withTheme(Button))>
        <ForwardRef(withTheme(Button))
          onPress={[Function]}
        >
          OK
        </ForwardRef(withTheme(Button))>
      </Dialog.Actions>
    </ForwardRef(withTheme(Dialog))>
  </Component>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Component
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    >
      <FlatList
        data={Array []}
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReachedThreshold={2}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        style={
          Object {
            "backgroundColor": "#bdbdbd",
            "margin": 0,
            "padding": 0,
            "width": "100%",
          }
        }
        updateCellsBatchingPeriod={50}
        windowSize={21}
      />
    </Component>
  </Component>
</Fragment>
`;
